### ISSUE TITLE
Failed to load tracks information:

### ISSUE THREAD

**Issue by charlescurley** [eneiluj/gpxpod#57]
I select a directory (under ~/nextcloud/ for maps, which has several known good GPX files (created by Navit from GPSD data). I get an error message, "Failed to load tracks information:", and no tracks. This is on Nextcloud 30.0.0. Database is sqlite 3.40.1-2 on Debian 12.

I also get the following sort of error in my nextcloud log.

Raw:

`{"reqId":"PN2IOGI0Eu76nMCKHCSn","level":3,"time":"2024-10-01T12:48:07-06:00","remoteAddr":"192.168.10.1","user":"charles","app":"index","method":"GET","url":"/index.php/apps/gpxpod/directories/1/tracks?directoryPath=%2FMaps&processAll=false","message":"Did not expect more than one result when executing: query \"SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)\"; ","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36","version":"30.0.0.14","exception":{"Exception":"OCP\\AppFramework\\Db\\MultipleObjectsReturnedException","Message":"Did not expect more than one result when executing: query \"SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)\"; ","Code":0,"Trace":[{"file":"/var/www/nextcloud/lib/public/AppFramework/Db/QBMapper.php","line":359,"function":"findOneQuery","class":"OCP\\AppFramework\\Db\\QBMapper","type":"->"},{"file":"/var/www/nextcloud/apps/gpxpod/lib/Db/DirectoryMapper.php","line":128,"function":"findEntity","class":"OCP\\AppFramework\\Db\\QBMapper","type":"->"},{"file":"/var/www/nextcloud/apps/gpxpod/lib/Controller/PageController.php","line":995,"function":"getDirectoryOfUserByPath","class":"OCA\\GpxPod\\Db\\DirectoryMapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":208,"function":"getTrackMarkersJson","class":"OCA\\GpxPod\\Controller\\PageController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":114,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":161,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":302,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1001,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":24,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/lib/public/AppFramework/Db/QBMapper.php","Line":270,"message":"Did not expect more than one result when executing: query \"SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)\"; ","exception":[],"CustomMessage":"Did not expect more than one result when executing: query \"SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)\"; "},"id":"66fc47754af60"}`

or, formatted:

`[index] Error: Did not expect more than one result when executing: query "SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)"; 
	GET /index.php/apps/gpxpod/directories/1/tracks?directoryPath=%2FMaps&processAll=false
	from 192.168.10.1 by charles at Oct 1, 2024, 12:48:07 PM
`

**Comment by julien-nc** [comment]
Can you execute this SQL query?
``` sql
SELECT * FROM oc_gpxpod_directories;
```

**Comment by charlescurley** [comment]
On Tue, 01 Oct 2024 14:42:19 -0700
Julien Veyssier ***@***.***> wrote:

> Can you execute this SQL query?
> ``` sql
> SELECT * FROM oc_gpxpod_directories;
> ```
> 

SQLite version 3.40.1 2022-12-28 14:03:47
Enter ".help" for usage hints.
sqlite> SELECT * FROM oc_gpxpod_directories;
3|charles|/Maps|0|0|1|0
4|charles|/Maps|0|0|1|0
sqlite> 

Hmmm, possibly from having installed, uninstalled, and then
re-installing. Except I was getting the error message after the first
installation, and the times on the log entries are consistent with the
first installation. And /Maps did show up twice during both
installations.

-- 
Does anybody read signatures any more?

https://charlescurley.com
https://charlescurley.com/blog/

**Comment by charlescurley** [comment]
On Tue, 1 Oct 2024 16:12:07 -0600
Charles Curley ***@***.***> wrote:

> SQLite version 3.40.1 2022-12-28 14:03:47
> Enter ".help" for usage hints.
> sqlite> SELECT * FROM oc_gpxpod_directories;  
> 3|charles|/Maps|0|0|1|0
> 4|charles|/Maps|0|0|1|0
> sqlite>   

Taking things a bit further. I verified that things had not changed as
above. I then disabled and then removed GpxPod. The same query gave the
same results.

After making proper backups (hurrah for sqlite putting everything in
one file!), I then listed and preserved the schema for GpxPod: ".sche
oc_gpx*". I then verified that there wasn't anything work preserving in
any of the 5 tables ("select * from oc_gpxpod_whatever;", etc.). I then
dropped those tables, and vacuumed the database.

That didn't work. I got:

Internal Server Error
The server was unable to complete your request.

If this happens again, please send the technical details below to the
server administrator.

More details can be found in the server log.

Technical details
Remote Address: 192.168.10.1
Request ID: 2ruc718pMR9uBbev0Ryz

and the log reports:

	
DbalException
An exception occurred while executing a query: SQLSTATE[HY000]: General
error: 1 no such table: oc_gpxpod_tracks

A subsequent SQL query showed that none of the five oc_gpx* tables were
re-created.

That was enough experimenting for one day. I then removed GpxPod, and
restored the database from its backup. Things are back to status quo
ante.


-- 
Does anybody read signatures any more?

https://charlescurley.com
https://charlescurley.com/blog/

**Comment by julien-nc** [comment]
Hey. I have no idea how the /Maps directory has been added twice. Anyway, there is a simple way forward: Just delete one of the 2 rows in oc_gpxpod_directories. This will fix the crash you initially reported.

About what you did: It was expected that it would break the app. Database changes (table creation/edition/deletion) are done with a "migration" system. The app contains migrations (the files in gpxpod/lib/Migration). When the app is installed or upgraded, Nextcloud runs the migrations that have not run yet. Nextcloud remembers which migrations have run in the oc_migrations table. When you manipulate the database structure manually, Nextcloud has now way to repair it. It won't run the GpxPod migrations again because they are all marked as done (they are in oc_migrations).
In this case, the way out would have been to run this query:
``` sql
DELETE FROM oc_migrations WHERE app="gpxpod";
```
and then to reinstall the app (or just disable and re-enable). This would have triggered all the migrations that would have created the tables again.

**Comment by julien-nc** [comment]
How did you add the directory? In the UI with "Add directories" -> "Add one directory"?

**Comment by charlescurley** [comment]
On Sat, 05 Oct 2024 14:25:39 -0700
Julien Veyssier ***@***.***> wrote:

> How did you add the directory? In the UI with "Add directories" ->
> "Add one directory"?

That's how I added the first one. I have no idea how the second one got
there.

-- 
Does anybody read signatures any more?

https://charlescurley.com
https://charlescurley.com/blog/

**Comment by julien-nc** [comment]
Any better after deleting one row?

**Comment by charlescurley** [comment]
On Sat, 05 Oct 2024 17:46:53 -0700
Julien Veyssier ***@***.***> wrote:

> Any better after deleting one row?
> 

No.

I get one Map directory, no track files in the display, and the
following error in the log:

{"reqId":"CH8x8mGaHTnVCjVaExwa","level":3,"time":"2024-10-05T19:47:49-06:00","remoteAddr":"192.168.10.1","user":"charles","app":"PHP","method":"GET","url":"/index.php/apps/gpxpod/maptiler/resources/logo.svg","message":"Array
to string conversion at
/var/www/nextcloud/lib/private/AppFramework/App.php#180","userAgent":"Mozilla/5.0
(X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)
Chrome/128.0.0.0
Safari/537.36","version":"30.0.0.14","data":{"app":"PHP"},"id":"6701ec772beef"}

[PHP] Error: Array to string conversion at
/var/www/nextcloud/lib/private/AppFramework/App.php#180 GET
/index.php/apps/gpxpod/maptiler/resources/logo.svg from 192.168.10.1 by
charles at Oct 5, 2024, 7:47:49 PM

However, when I go from the log back to the app, I still only see one
Maps directory.

This time, just for the halibut, I clicked on that directory. My three
.gpx files showed up. Oh, OK, even if there is only one directory I
have to click on it. User ignorance.

Nice app, once I figure out how to use it. Thanks.

-- 
Does anybody read signatures any more?

https://charlescurley.com
https://charlescurley.com/blog/

**Comment by julien-nc** [comment]
The warning you see in the logs is fixed in the development version. It will come with the next release.

I'll check if selecting the first added directory can be done easily.

Please reopen this if you face the initial issue again. It would be helpful to know which sequence of action led to having the same directory added twice in the DB.

**Comment by creideiki** [comment]
I can reliably trigger this issue immediately upon adding a directory of GPX files through the new web UI. GPXPod 7.0.1 on NextCloud 30.0.0, PHP 8.3.12 and PostgreSQL 17.0. I have tried several times to clear the database and re-add directories, and get duplicate database entries each time.

I keep my GPX files in a tree structure by year and month:
```
$ tree -d ~/Nextcloud/GPX/2024/
/home/creideiki/Nextcloud/GPX/2024/
├── 2024-02
├── 2024-03
├── 2024-04
├── 2024-05
├── 2024-06
├── 2024-07
├── 2024-08
├── 2024-09
└── 2024-10
```
To start with, the ones up to and including `2024/2024-09/` were added in the old UI, and those worked in the new web UI also. When I added `2024/2024-10/` through the new UI, I had this problem on that directory only - it appeared twice in the list of directories in the web UI and the database, and attempting to enter it through the web UI gave me a pop-up with "Failed to load tracks information:".

To try to remedy this problem, I cleared the database:
```sql
delete from gpxpod_directories;
```
And re-added the root directory through the web UI:

GPXPod -> Add directories -> Recursively add a directory -> GPX -> Choose GPX

Immediately after adding the root directory, every directory appeared once in the web UI.

Without clicking anything else in the web UI, I looked at the database, and saw that every directory was already listed twice:
```sql
nextcloud=# select * from gpxpod_directories where path like '%2024%';
 id  |   user    |       path        | is_open | sort_order | sort_ascending | display_recursive 
-----+-----------+-------------------+---------+------------+----------------+-------------------
 201 | creideiki | /GPX/2024/2024-02 |       0 |          0 |              1 |                 0
 202 | creideiki | /GPX/2024/2024-02 |       0 |          0 |              1 |                 0
 203 | creideiki | /GPX/2024/2024-03 |       0 |          0 |              1 |                 0
 204 | creideiki | /GPX/2024/2024-03 |       0 |          0 |              1 |                 0
 205 | creideiki | /GPX/2024/2024-04 |       0 |          0 |              1 |                 0
 206 | creideiki | /GPX/2024/2024-04 |       0 |          0 |              1 |                 0
 207 | creideiki | /GPX/2024/2024-05 |       0 |          0 |              1 |                 0
 208 | creideiki | /GPX/2024/2024-05 |       0 |          0 |              1 |                 0
 209 | creideiki | /GPX/2024/2024-06 |       0 |          0 |              1 |                 0
 210 | creideiki | /GPX/2024/2024-06 |       0 |          0 |              1 |                 0
 211 | creideiki | /GPX/2024/2024-07 |       0 |          0 |              1 |                 0
 212 | creideiki | /GPX/2024/2024-07 |       0 |          0 |              1 |                 0
 213 | creideiki | /GPX/2024/2024-08 |       0 |          0 |              1 |                 0
 214 | creideiki | /GPX/2024/2024-08 |       0 |          0 |              1 |                 0
 215 | creideiki | /GPX/2024/2024-09 |       0 |          0 |              1 |                 0
 216 | creideiki | /GPX/2024/2024-09 |       0 |          0 |              1 |                 0
 217 | creideiki | /GPX/2024/2024-10 |       0 |          0 |              1 |                 0
 218 | creideiki | /GPX/2024/2024-10 |       0 |          0 |              1 |                 0
(18 rows)
```
I reloaded the GPXPod web UI. Every directory was now listed twice, and clicking on any of them gave the same pop-up with "Failed to load tracks information:" and a 500 Internal Server Error in the JavaScript logs.

The Nextcloud server logs said "Did not expect more than one result":

```json
{
  "reqId": "E9rPV8mak89QrJBkluOp",
  "level": 3,
  "time": "2024-10-19T14:21:13+02:00",
  "remoteAddr": "REDACTED",
  "user": "creideiki",
  "app": "index",
  "method": "GET",
  "url": "/nextcloud/index.php/apps/gpxpod/directories/203/tracks?directoryPath=%2FGPX%2F2024%2F2024-03&processAll=false",
  "message": "Did not expect more than one result when executing: query \"SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)\"; ",
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0",
  "version": "30.0.0.14",
  "exception": {
    "Exception": "OCP\\AppFramework\\Db\\MultipleObjectsReturnedException",
    "Message": "Did not expect more than one result when executing: query \"SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)\"; ",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/localhost/htdocs/nextcloud/lib/public/AppFramework/Db/QBMapper.php",
        "line": 359,
        "function": "findOneQuery",
        "class": "OCP\\AppFramework\\Db\\QBMapper",
        "type": "->"
      },
      {
        "file": "/var/www/localhost/htdocs/nextcloud/apps/gpxpod/lib/Db/DirectoryMapper.php",
        "line": 128,
        "function": "findEntity",
        "class": "OCP\\AppFramework\\Db\\QBMapper",
        "type": "->"
      },
      {
        "file": "/var/www/localhost/htdocs/nextcloud/apps/gpxpod/lib/Controller/PageController.php",
        "line": 1001,
        "function": "getDirectoryOfUserByPath",
        "class": "OCA\\GpxPod\\Db\\DirectoryMapper",
        "type": "->"
      },
      {
        "file": "/var/www/localhost/htdocs/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
        "line": 208,
        "function": "getTrackMarkersJson",
        "class": "OCA\\GpxPod\\Controller\\PageController",
        "type": "->"
      },
      {
        "file": "/var/www/localhost/htdocs/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
        "line": 114,
        "function": "executeController",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->"
      },
      {
        "file": "/var/www/localhost/htdocs/nextcloud/lib/private/AppFramework/App.php",
        "line": 161,
        "function": "dispatch",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->"
      },
      {
        "file": "/var/www/localhost/htdocs/nextcloud/lib/private/Route/Router.php",
        "line": 302,
        "function": "main",
        "class": "OC\\AppFramework\\App",
        "type": "::"
      },
      {
        "file": "/var/www/localhost/htdocs/nextcloud/lib/base.php",
        "line": 1001,
        "function": "match",
        "class": "OC\\Route\\Router",
        "type": "->"
      },
      {
        "file": "/var/www/localhost/htdocs/nextcloud/index.php",
        "line": 24,
        "function": "handleRequest",
        "class": "OC",
        "type": "::"
      }
    ],
    "File": "/var/www/localhost/htdocs/nextcloud/lib/public/AppFramework/Db/QBMapper.php",
    "Line": 270,
    "message": "Did not expect more than one result when executing: query \"SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)\"; ",
    "exception": {},
    "CustomMessage": "Did not expect more than one result when executing: query \"SELECT * FROM `*PREFIX*gpxpod_directories` WHERE (`path` = :dcValue1) AND (`user` = :dcValue2)\"; "
  }
}
```
Adding just a single directory instead of the entire tree recursively also creates duplicates:
```sql
nextcloud=# select * from gpxpod_directories;
 id  |   user    |       path        | is_open | sort_order | sort_ascending | display_recursive 
-----+-----------+-------------------+---------+------------+----------------+-------------------
 375 | creideiki | /GPX/2024/2024-10 |       0 |          0 |              1 |                 0
 376 | creideiki | /GPX/2024/2024-10 |       0 |          0 |              1 |                 0
(2 rows)
```
If I remove the duplicates, either by running raw SQL queries or through the web UI by clicking ... -> Remove from list, the remaining directory works in the web UI.

**Comment by DerTee** [comment]
This issue is still happening in GPXPod 7.0.1 on my Nextcould 30.0.1, PHP 8.1.29

I had the same issue and I simply fixed it ad-hoc in the database by removing one id, so the gpxpod is working for me for now. So a simple ```delete from gpxpod_directories where id=ANY_DUPLICATE_ID``` is sufficient for a quick fix, if people know a little DB stuff.

(Sorry if this is obvious to you!!) As a first partial fix I would suggest ensuring that the database does not allow invalid state, i.e. set a Unique Constraint on the table gpxpod_directories for the fields `user` and `name`. This will lead to an error at the point where the actual problem sits. From there it should be easier to track down the root cause, plus, if it ever happens again on the PHP side, you get a better error message to begin with.

**Comment by creideiki** [comment]
TL;DR: lightly edited stream-of-consciousness debugging session, suggesting that the problem lies in client-side JavaScript which tries to add each directory twice. And I'm not going to dig into client-side JavaScript. Also, `lib/Db/DirectoryMapper.php:createDirectory()` has a sanity check which doesn't work, and should be moved to the database.

> As a first partial fix I would suggest ensuring that the database does not allow invalid state, i.e. set a Unique Constraint on the table gpxpod_directories for the fields `user` and `name`. This will lead to an error at the point where the actual problem sits.

I just tried that, and it's unfortunately not that simple. I emptied the table and added the constraint:
```sql
delete from gpxpod_directories;
alter table gpxpod_directories add constraint gpxpod_unique_directories unique (path);
```
and then used the web UI to add a single directory. The web browser console showed an error 400:
`response: Object { data: 'Impossible to insert. An exception occurred while executing a query: SQLSTATE[23505]: Unique violation: 7 ERROR:  duplicate key value violates unique constraint "gpxpod_unique_directories"\nDETAIL:  Key (path)=(/GPX/2024/2024-10) already exists.', status: 400, statusText: "Bad request", … }`
but the stack trace only pointed to `onDirectoryAdd` in minified client-side JavaScript code.

Neither the PostgreSQL, Apache nor NextCloud logs contained any error message.

That "Impossible to insert" message is only generated if `lib/Db/DirectoryMapper.php:createDirectory()` fails, but that function already contains a check for not creating duplicate entries, so a duplicate call should never even make it to the database.
https://github.com/julien-nc/gpxpod/blob/d70024069a6d5599b2bde6e4bcd5b167b7858df6/lib/Db/DirectoryMapper.php#L163-L188

Instead, I called upon my very best copy-and-pasting skills to edit PHP for the very first time and catch any errors returned by `$this->insert($dir);` and log a stack trace:
```php
                error_log(sprintf("In createDirectory(%s, %s)", $path, $user));
                try {
                        $createdDirectory = $this->insert($dir);
                } catch (Exception $ex) {
                        error_log(json_encode(debug_backtrace(0)));
                }
```
Unfortunately, all this says is that the error comes the second time `createDirectory()` is called, which comes from  an `addDirectory` action:
```
[Sun Oct 20 12:43:06.186153 2024] [pid 23528:tid 23528] [client 192.168.0.17:34090] In createDirectory(/GPX/2024/2024-10, creideiki)
[Sun Oct 20 12:43:06.186206 2024] [pid 23605:tid 23605] [client 192.168.0.17:34054] In createDirectory(/GPX/2024/2024-10, creideiki)
[Sun Oct 20 12:43:06.190447 2024] [pid 23605:tid 23605] [client 192.168.0.17:34054] [{"file":"\\/var\\/www\\/localhost\\/htdocs\\/nextcloud\\/apps\\/ ...
```
```json
[
  {
    "file": "/var/www/localhost/htdocs/nextcloud/apps/gpxpod/lib/Controller/PageController.php",
    "line": 743,
    "function": "createDirectory",
    "class": "OCAGpxPodDbDirectoryMapper",
    "type": "->",
    "args": [
      "/GPX/2024/2024-10",
      "creideiki",
      false
    ]
  },
  {
    "file": "/var/www/localhost/htdocs/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
    "line": 208,
    "function": "addDirectory",
    "class": "OCAGpxPodControllerPageController",
    "type": "->",
    "args": [
      "/GPX/2024/2024-10",
      false
    ]
  },
  {
    "file": "/var/www/localhost/htdocs/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
    "line": 114,
    "function": "executeController",
    "class": "OCAppFrameworkHttpDispatcher",
    "type": "->",
    "args": [
      {},
      "addDirectory"
    ]
  },
  {
    "file": "/var/www/localhost/htdocs/nextcloud/lib/private/AppFramework/App.php",
    "line": 161,
    "function": "dispatch",
    "class": "OCAppFrameworkHttpDispatcher",
    "type": "->",
    "args": [
      {},
      "addDirectory"
    ]
  },
  {
    "file": "/var/www/localhost/htdocs/nextcloud/lib/private/Route/Router.php",
    "line": 302,
    "function": "main",
    "class": "OCAppFrameworkApp",
    "type": "::",
    "args": [
      "OCAGpxPodControllerPageController",
      "addDirectory",
      {},
      {
        "_route": "gpxpod.page.adddirectory"
      }
    ]
  },
  {
    "file": "/var/www/localhost/htdocs/nextcloud/lib/base.php",
    "line": 1001,
    "function": "match",
    "class": "OCRouteRouter",
    "type": "->",
    "args": [
      "/apps/gpxpod/directories"
    ]
  },
  {
    "file": "/var/www/localhost/htdocs/nextcloud/index.php",
    "line": 24,
    "function": "handleRequest",
    "class": "OC",
    "type": "::",
    "args": []
  }
]
```
One interesting thing is that the PHP code for catching duplicates has apparently not triggered:
```php
 	try { 
 		// do not create if one with same path/userId already exists 
 		$dir = $this->getDirectoryOfUserByPath($path, $user); 
 		throw new Exception('Already exists'); 
 	}
```
because if it had, the code I inserted to catch and log errors would not have run.

Another interesting thing is that the two calls to `createDirectory()` are run in two different PHP interpreters, less than half a millisecond apart. Could we be looking at a race condition, where the client-side JavaScript sends two calls, which race in the PHP code but are serialized in the database?

Let's break out Firefox's network debugger.

When clicking the "Choose $DIR" button in the web UI's directory picker, two XMLHTTPRequests are fired with the exact same contents:

```
POST /nextcloud/index.php/apps/gpxpod/directories HTTP/1.1
Host: REDACTED
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0
Accept: application/json, text/plain, */*
Accept-Language: en-GB,en-US;q=0.8,en;q=0.6,sv-SE;q=0.4,sv;q=0.2
Accept-Encoding: gzip, deflate, br, zstd
Content-Type: application/json
Content-Length: 28
X-Requested-With: XMLHttpRequest, XMLHttpRequest
requesttoken: REDACTED
Origin: REDACTED
DNT: 1
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
Connection: keep-alive
Cookie: REDACTED
Priority: u=0

{"path":"/GPX/2024/2024-10"}
```

The first response has the status "200 OK" and the contents is the ID of the newly inserted row in the database. The second response has a status "500 Internal Server Error" and the contents contain a NextCloud server log ID, which notes that my edited function returns `null` instead of a directory.

Which leads me to believe that what happens is:
1. The front-end JavaScript code sends two requests simultaneously
2. The two requests get handled by two different PHP interpreters in the web server
3. Both requests get past the PHP code checking for duplicates
4. Both requests get to the database simultaneously
   1. In the upstream code, the requests do not conflict, and both are added to the table
   2. With the unique constraint added, one of the requests is rejected by the database

As for why two XMLHTTPRequests are sent simultaneously, I leave that to someone willing to dig into client-side JavaScript.

In any case, the PHP code which checks for duplicate entries is an obvious time-of-check-to-time-of-use violation, because things can happen in the database between the check and the insertion. The PHP check should be removed, the constraint moved to the database, and database errors handled by the PHP code.

**Comment by julien-nc** [comment]
I was finally able to reproduce the issue and partially understand the cause. The only place I can reproduce is a production instance. No idea why.
The symptom is that when the file picker is opened, Vite (the JS bundler) dynamically loads the main script again, which mounts the main component a second time so every event sent in the event bus is caught twice, once by each App instance.

A small change was enough to dodge this: Only mount the main component on the "DOMContentLoaded" event. I would prefer understanding why Vite loads the main script again but at least this issue should be fixed now.
v7.0.2 is out and includes the fix. Could you confirm?

I also added the unique constraint on the user and path Db columns.

**Comment by creideiki** [comment]
I have upgraded to 7.0.2, and I'm now seeing only one XMLHTTPRequest when adding a directory, and the directory list contains no duplicates. It seems to be fixed.

**Comment by DerTee** [comment]
Same for me after the upgrade to 7.0.2: Only one XMLHTTPRequest and only one added database entry -> now it works as expected, thank you!

**Comment by charlescurley** [comment]
That's better. I can add directories.

However, now I get a different issue. Nothing in the log, but I see a banner over the GpxPod browser tab:

```Failed to add directory: Impossible to insert. An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: oc_gpxpod_directories.user, oc_gpxpod_directories.path```

I now have two directories. I see three files in each one, which is correct. But I don't see the tracks on the map or the endpoint markers as before, even after the track is digested and the track entry in the directory is given a blue box.

Other people haven't reported this sort of issue. I am using. I wonder if using sqlite has anything to do with it.

**Comment by julien-nc** [comment]
@charlescurley If you are using GpxPod v7.0.2, the network errors should not be displayed in the page anymore. Try to "hard" refresh (usually CTRL+F5). Does it change anything?

If not:
Do you have rows in oc_gpxpod_directories that have the same "user" and "path"? If so, what if you delete them? Can you then add directories without seeing the error again?

**Comment by charlescurley** [comment]
Argghh.  I have shut down and restarted the web server and the browser.

I deleted (from the web client) all directories. First thing to note is that when I deleted the last one, it was deleted from the database but the display did not refresh, so it looked like the directory hadn't been deleted. A second attempt to delete it failed.

I then tried to add a new directory. The display claims "No files in here" which is false; there are three gpx files in there.

I choose the directory anyway, got "Failed to add directory" on the browser tab. I refreshed the tab, and the directory showed up. I then ran the following query:

```
root@hawk:/var/local/nc.data# sqlite3 --line nextcloud.db 'select * from oc_gpxpod_directories;'
               id = 8
             user = charles
             path = /Maps/2023.11.24
          is_open = 1
       sort_order = 0
   sort_ascending = 1
display_recursive = 0
root@hawk:/var/local/nc.data# 
```
It is now picking up and digesting the tracks therein.

I then added another directory. I got:

```
Failed to add directory: Impossible to insert. An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: oc_gpxpod_directories.user, oc_gpxpod_directories.path
```
but the directory showed up. The above query now shows two rows in the database.

Further question: If I select one directory, then another, what should happen? Should the first directory's tracks show up, then when I select the second directory, do the first directory's tracks go away, or do they stay?

**Comment by charlescurley** [comment]
For what it's worth, this now seems to be working correctly. GpxPod 7.0.4, Nextcloud Hub 9 (30.0.5). Thank you.

As far as I am concerned, you are welcome to close this.
